posts/2021-03-25-Rust-1.51.0.md

```diff
@@ -0,0 +1,211 @@
+
+While this can lead to some crates compiling more than once, this should provide a much more intuitive development experience when using features with cargo. If you'd like to know more, you can also read the ["Feature Resolver"][feature-resolver@2.0] section in the Cargo Book for more information. We'd like to thank the cargo team and everyone involved for all their hard work in designing and implementing this new feature (resolver)!
+
+```toml
+[package]
+resolver = "2"
+# Or if you're using a workspace
+[workspace]
+resolver = "2"
+```
+
+[rfc 2957]: https://rust-lang.github.io/rfcs/2957-cargo-features2.html
+[feature-resolver@2.0]: https://doc.rust-lang.org/nightly/cargo/reference/features.html#feature-resolver-version-2
+
+### Splitting Debug Information
+While not often highlighted in the release, the Rust teams are constantly working on improving Rust's compile times, and this release marks one of the biggest improvements in a long time for Rust on macOS platforms. Debug information is code that maps the binary code back to your source code, so that the program can give you more information about what went wrong at run time. In macOS this is done with a tool called `dsymutil` which collects all the debug information into a single `.dSYM` folder.
+
+This tool provides a lot of benefits, however, one of the drawbacks has been that `dsymutil` is not really compatible with incremental recompilation. This means that even when you make a small change such as a function name, `dsymutil` will need to run over the entire final binary to produce the final `.dSYM` folder. This can sometimes add a lot to the build time, (especially for larger projects), but this has been a necessary step as without it Rust wouldn't be able to parse debug info on macOS.
+
+With the implementation of the Rust-based `backtrace` crate backend, we can now parse debuginfo even without running `dsymutil`. This can significantly speed up builds that include debuginfo and significantly reduce the amount of disk space used. We haven't run extensive benchmarks but have seen a lot of anecdata of people's builds building a lot faster on macOS with this behaviour.
+
+You can enable this new behaviour by setting the `-Csplit-debuginfo=unpacked` flag when running `rustc`, or by setting the [`split-debuginfo`] `[profile]` option to `unpacked` in Cargo. The "unpacked" option instructs rustc to leave the .o object files in the build output directory instead of deleting them, and skips the step of running dsymutil. Rust's backtrace support is smart enough to know how to find these .o files. Tools such as lldb also know how to do this. This should work as long as you don't need to move the binary to a different location while retaining the debug information.
+
+```toml
+[profile.dev]
+split-debuginfo = "unpacked"
```

Comment:
We should tell people that they likely want to put this in ~/.cargo/config I think, not Cargo.toml and only on macOS, right? On other platforms this will lead to an error indicating that unpacked behavior is currently unstable I suspect.

Reply:
It is ignored on other platforms, so it shouldn't cause an error.

It wouldn't hurt to suggest using a config file.  I'm not sure if this blog post is getting a little too long, though.

Here's a hastily written suggestion:

> You can place the profile definition in a local Cargo config file `.cargo/config.toml` if you don't want to include it in the `Cargo.toml` shared with everyone. The `split-debuginfo` setting will be ignored on platforms where it is not supported.

Reply:
Ah, ok, if it's ignored seems good - I think I may have been confused as rustc claims it will error.

Reply:
Ah, yes, the ignore behavior is cargo-specific (it probes if it is supported).  `rustc` itself returns an error if not supported.